File: py/engit/_cleanup.py
"""engit cleanup command — tidy up merged and stale local branches.

Performs the following cleanup steps:

* Runs ``git remote prune <remote>`` to remove stale remote-tracking refs.
* Deletes local branches that have been merged into the current branch.
* Deletes local branches whose remote-tracking branch no longer exists.

The current branch and remote default branches (``main``, ``master``,
``develop``) are always protected.
"""

from __future__ import annotations

from pathlib import Path

from ._exceptions import GitError
from ._git import (
    require_git_repo,
    get_current_branch,
    get_merged_branches,
    get_remote_tracking_branch,
    delete_local_branch,
    prune_remote,
    _run,
)

# Branch names that should never be force-deleted by cleanup.
_PROTECTED = frozenset({'main', 'master', 'develop'})


def run_cleanup(
    *,
    remote: str = 'origin',
    noop: bool = False,
    cwd: Path | None = None,
) -> None:
    """Clean up merged and stale local branches.

    Args:
        remote: Remote name to prune and check. Defaults to ``'origin'``.
        noop: When ``True``, print what would happen without deleting anything.
        cwd: Git working directory. Defaults to the current directory.

    Raises:
        ~._exceptions.NotAGitRepoError: If not inside a git repo.

    """
    require_git_repo(cwd=cwd)

    current = get_current_branch(cwd=cwd) or ''

    # ---- Prune stale remote-tracking refs ----
    try:
        if noop:
            _run('remote', 'prune', remote, '--dry-run', cwd=cwd)
        else:
            prune_remote(remote, cwd=cwd)
            print(f'Pruned stale remote-tracking refs for {remote}.')
    except GitError as exc:
        print(f'Warning: could not prune remote "{remote}": {exc}')

    deleted: list[str] = []
    skipped: list[str] = []

    # ---- Delete merged branches ----
    for branch in get_merged_branches(cwd=cwd):
        if branch == current or branch in _PROTECTED:
            continue
        if noop:
            print(f'  Would delete {branch} [merged]')
            continue
        try:
            delete_local_branch(branch, force=False, cwd=cwd)
            deleted.append(branch)
            print(f'  Deleted {branch} [merged]')
        except GitError as exc:
            skipped.append(branch)
            print(f'  Skipped {branch}: {exc}')

    # ---- Delete branches whose remote is gone ----
    try:
        raw = _run('branch', cwd=cwd)
    except GitError:
        raw = ''

    for line in raw.splitlines():
        branch = line.lstrip('* ').strip()
        if not branch or branch == current or branch in _PROTECTED:
            continue
        if branch in deleted:
            continue

        tracking = get_remote_tracking_branch(branch, cwd=cwd)
        if tracking is None:
            continue  # No upstream set; leave it alone.

        # Check whether the remote branch still exists
        try:
            _run('show-ref', '--verify', '--quiet', f'refs/remotes/{tracking}', cwd=cwd)
            # Still exists — skip
        except GitError:
            # Remote branch is gone
            if noop:
                print(f'  Would delete {branch} [remote deleted]')
                continue
            try:
                delete_local_branch(branch, force=True, cwd=cwd)
                deleted.append(branch)
                print(f'  Deleted {branch} [remote deleted]')
            except GitError as exc:
                skipped.append(branch)
                print(f'  Skipped {branch}: {exc}')

    if not noop:
        if deleted:
            print(f'\nDeleted {len(deleted)} branch(es).')
            print('Use `git branch <name> <revision>` or `git reflog` to restore if needed.')
        else:
            print('No local branches to clean up.')
        if skipped:
            print(f'Warning: {len(skipped)} branch(es) could not be deleted: {", ".join(skipped)}')

File: rust/engit-core/src/cleanup.rs
//! Merged / stale branch cleanup helpers.

use std::path::Path;

use crate::git::{
    delete_local_branch, get_current_branch, get_merged_branches, get_remote_tracking_branch,
    prune_remote, require_git_repo, run_git,
};

/// Branch names never deleted by cleanup.
pub const PROTECTED_BRANCHES: &[&str] = &["main", "master", "develop"];

/// Clean up merged and stale local branches.
pub fn run_cleanup(remote: &str, noop: bool, cwd: Option<&Path>) -> crate::error::Result<()> {
    require_git_repo(cwd)?;

    let current = get_current_branch(cwd).unwrap_or_default();

    if noop {
        let _ = run_git(["remote", "prune", remote, "--dry-run"], cwd);
    } else if let Err(error) = prune_remote(remote, cwd) {
        println!("Warning: could not prune remote \"{remote}\": {error}");
    } else {
        println!("Pruned stale remote-tracking refs for {remote}.");
    }

    let mut deleted = Vec::new();
    let mut skipped = Vec::new();

    for branch in get_merged_branches(cwd) {
        if branch == current || PROTECTED_BRANCHES.contains(&branch.as_str()) {
            continue;
        }
        if noop {
            println!("  Would delete {branch} [merged]");
            continue;
        }
        match delete_local_branch(&branch, false, cwd) {
            Ok(()) => {
                deleted.push(branch.clone());
                println!("  Deleted {branch} [merged]");
            }
            Err(error) => {
                skipped.push(branch.clone());
                println!("  Skipped {branch}: {error}");
            }
        }
    }

    let raw = run_git(["branch"], cwd).unwrap_or_default();
    for line in raw.lines() {
        let branch = line.trim_start_matches('*').trim();
        if branch.is_empty()
            || branch == current
            || PROTECTED_BRANCHES.contains(&branch)
            || deleted
                .iter()
                .any(|deleted_branch| deleted_branch == branch)
        {
            continue;
        }

        let Some(tracking) = get_remote_tracking_branch(branch, cwd) else {
            continue;
        };
        let ref_name = format!("refs/remotes/{tracking}");
        if run_git(["show-ref", "--verify", "--quiet", &ref_name], cwd).is_ok() {
            continue;
        }

        if noop {
            println!("  Would delete {branch} [remote deleted]");
            continue;
        }

        match delete_local_branch(branch, true, cwd) {
            Ok(()) => {
                deleted.push(branch.to_string());
                println!("  Deleted {branch} [remote deleted]");
            }
            Err(error) => {
                skipped.push(branch.to_string());
                println!("  Skipped {branch}: {error}");
            }
        }
    }

    if !noop {
        if deleted.is_empty() {
            println!("No local branches to clean up.");
        } else {
            println!("\nDeleted {} branch(es).", deleted.len());
            println!("Use `git branch <name> <revision>` or `git reflog` to restore if needed.");
        }
        if !skipped.is_empty() {
            println!(
                "Warning: {} branch(es) could not be deleted: {}",
                skipped.len(),
                skipped.join(", ")
            );
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::PROTECTED_BRANCHES;

    #[test]
    fn protected_branch_list_matches_python_defaults() {
        assert_eq!(PROTECTED_BRANCHES, &["main", "master", "develop"]);
    }
}
